### PR TITLE
Add RouteHandlerContext as parameter to IRouteHandlerFilter.InvokeAsync

### LINKED ISSUE #40724: Support accessing `RouteHandlerContext` in `IRouteHandlerFilter`
## Background and Motivation

The `IRouteHandlerFilter` implementations provides support for accessing the `RouteHandlerInvocationContext` for accessing information about the current request invocation.

In #40513 , we added support for a filter factory API that allowed users to provide a `Func<RouteHandlerContext, RouteHandlerFilterDelegate>` that allowed them to construct a filter using information available in the `RouteHandlerContext` such as `MethodInfo` and `EndpointMetadata`.

The same functionality doesn't exist for users who are implementing their filters using the `IRouteHandlerFilter` and injecting them into DI. To that end, we want to be able to provide the `RouteHandlerContext` to the user's `IRouteHandlerFilter` implementation.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http;

public interface IRouteHandlerFilter
{
-    ValueTask<object?> InvokeAsync(RouteHandlerInvocationContext context, RouteHandlerFilterDelegate next);
+    ValueTask<object?> InvokeAsync(RouteHandlerInvocationContext context, RouteHandlerFilterDelegate next, RouteHandlerContext routeHandlerContext);
}
```

## Usage Examples

```csharp
class IncrementArgFilter : IRouteHandlerFilter
{
    public async ValueTask<object?> InvokeAsync(RouteHandlerInvocationContext context, RouteHandlerFilterDelegate next, RouteHandlerContext routeHandlerContext)
    {
        if (routeHandlerContext.MethodInfo.GetParameters().Length == 1
            && routeHandlerContext.MethodInfo.GetParameters()[0].ParameterType == typeof(int))
        {
            context.Parameters[0] = ((int)context.Parameters[0]!) + 1;
        }

        return await next(context);
    }
}
```

## Alternative Designs

Some of the alternatives considered include:

- Setting the `RouteHandlerContext` as an argument to the `ObjectFactory` used to activate the filter. However, we need to make sure this doesn't break for scenarios where the user doesn't use it
- Setting the `RouteHandlerContext` as a property on the `IRouteHandlerFilter` and then setting it in the code gen/filter construction

## Risks

The `InvokeAsync` method takes in both a `RouteHandlerContext` and a `RouteHandlerInvocationContext` which might be confusing to some users and require education on distinguishing between the two.

### PR DESCRIPTION
**Pull request by captainsafia**
Fixes https://github.com/dotnet/aspnetcore/issues/40724

### CODE REVIEW

**Comment by davidfowl** [comment]
I'm not a huge fan of this. It means that if we ever want to support a lifetime other than instantiate per request, then there's no way to cache the data from this context. That's why I prefer the constructor approach.

File: src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs [hunk 1]
```diff
@@ -32,11 +32,22 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
...
+        catch (InvalidOperationException)
+        {
+            filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), Type.EmptyTypes);
+        }
+
         builder.RouteHandlerFilterFactories.Add((routeHandlerContext, next) => (context) =>
         {
-            var filter = (IRouteHandlerFilter)filterFactory.Invoke(context.HttpContext.RequestServices, Array.Empty<object>());
-            return filter.InvokeAsync(context, next, routeHandlerContext);
+            var filter = (IRouteHandlerFilter)filterFactory.Invoke(context.HttpContext.RequestServices, new[] { routeHandlerContext });
```
Comment (BrennanConroy) [inline src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs:49]: Is this is called per request? If so the `new[] { routeHandlerContext }` should be moved into the out lambda.

**Review by BrennanConroy — APPROVED** [review]
Does any of this new functionality have docs? And if not have we applied the doc label to the issues/prs?

**Comment by captainsafia** [comment]
> Does any of this new functionality have docs? And if not have we applied the doc label to the issues/prs?

Docs issue is https://github.com/dotnet/AspNetCore.Docs/issues/25483.

Will do a follow-up to update runtime with nicer APIs for registering optional arguments for `ActivatorUtilities` but merging this for now.

**Comment by davidfowl** [comment]
Left a comment, I really don’t like this exception but if we have a follow up issue it’s fine. We avoid first chance exceptions like this in the overall stack

File: src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs [hunk 2]
```diff
@@ -32,11 +32,26 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
...
-        var filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), Type.EmptyTypes);
-        builder.RouteHandlerFilterFactories.Add((routeHandlerContext, next) => (context) =>
+        // We call `CreateFactory` twice here since the `CreateFactory` API does not support optional arguments.
+        // See https://github.com/dotnet/runtime/issues/67309 for more info.
+        ObjectFactory filterFactory;
+        try
         {
-            var filter = (IRouteHandlerFilter)filterFactory.Invoke(context.HttpContext.RequestServices, Array.Empty<object>());
-            return filter.InvokeAsync(context, next);
+            filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), new[] { typeof(RouteHandlerContext) });
```
Comment (davidfowl) [inline src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs:40]: Could we manually check to see if there are any constructors that take `RouteHandlerContext` before doing this?

**Pull request merged**